src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -45,5 +50,386 @@ protected override void InsertItem(int index, IDeltaSetItem item)
+
+            Debug.Assert(apiHandlerOfT != null, "apiHandlerOfT != null");
+
+            DeltaSet<TStructuralType> deltaSet = CreateDeltaSet();
+
+            foreach (Delta<TStructuralType> changedObj in Items)
+            {
+                ODataAPIHandler<TStructuralType> handler = apiHandlerOfT;
+                bool hasODataId = false;
+
+                if (apiHandlerFactory != null)
+                {
+                    //{
+                    //    "@odata.context":"http://localhost:6285/odata/$metadata#Customers/$delta",
+                    //    "value":[
+                    //        {
+                    //            "@odata.type":"CustomersApiEF.Models.Customer",
+                    //            "Id":1,
+                    //            "Name":"Customer1",
+                    //            "Orders@odata.delta":[
+                    //                {
+                    //                    "@odata.id":"Customers(3)/Orders(1005)",
+                    //                    "Quantity": 1005
+                    //                },
+                    //                {
+                    //                    "Id": 2000,
+                    //                    "Price":200,
+                    //                    "Quantity":90
+                    //                }
+                    //            ]
+                    //        }
+                    //    ]
+                    //}
+
+                    // If we have a request payload above and we are handling the changed values in Orders,
+                    // The apiHandlerOfT is OrdersAPIHandler with parent as Customers(1).
+                    // The object with "@odata.id" will have OrdersAPIHandler with parent as Customers(3).
+                    // The object with id 2000 will have OrdersAPIHandler similar to apiHandlerOfT.
+                    // The codebelow ensures we use the correct handler.
+                    // NOTE: Handler when we have odata.id will always be different than the parent handler.
+                    // However this is true if handlers are applied correctly.
+
+                    ODataAPIHandler<TStructuralType> odataPathApiHandler = apiHandlerFactory.GetHandler(changedObj.ODataPath) as ODataAPIHandler<TStructuralType>;
+
+                    if (odataPathApiHandler != null && changedObj.ODataPath.Any())
+                    {
+                        // Using a 3rd party library to compare objects
+                        // https://github.com/GregFinzer/Compare-Net-Objects/wiki/Getting-Started
+                        // TODO:
+                        // a) Confirm to perf regressions are caused byt his library.
+                        // b) Investigate if there is a better library.
+
+                        KellermanSoftware.CompareNetObjects.CompareLogic compare = new KellermanSoftware.CompareNetObjects.CompareLogic();
```

Comment:
why do we need to use this library here?

Reply:
To compare 2 objects, in this case two handlers.

Reply:
Like there is no other way of comparing the objects without using a 3rd party library?

Reply:
There is no straight forward way without implementing our own `IComparer`.